pkg/controller/zerocapacity/zero_capacity.go

```diff
@@ -0,0 +1,452 @@
+	resource := reflect.TypeOf(reconciler.Type()).Name()
+	namespace := request.Namespace
+
+	reqLogger := z.Log.WithValues("Request.Namespace", namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling " + resource)
+	defer reqLogger.Info("----- End Reconciling " + resource)
+
+	instance, err := reconciler.ResourceInstance(request.NamespacedName, z)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch %s CR '%s': %w", resource, request.Name, err)
+	}
+
+	phase := instance.Phase()
+	switch phase {
+	case "":
+		return z.initializeResources(request, instance)
+	case ZeroSucceeded, ZeroFailed:
+		return z.cleanupResources(request)
+	}
+
+	infinispan := &v1.Infinispan{}
+	clusterName := instance.Spec().Cluster
+	clusterObjKey := client.ObjectKey{
+		Namespace: namespace,
+		Name:      clusterName,
+	}
+	if err := z.Get(context.Background(), clusterObjKey, infinispan); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("CR '%s' not found: %w", clusterName, err)
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch CR '%s': %w", clusterName, err)
+	}
+
+	user := consts.DefaultOperatorUser
+	pass, err := users.PasswordFromSecret(user, infinispan.GetSecretName(), namespace, z.Kube)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	httpConfig := ispnclient.HttpConfig{
+		Username:  user,
+		Password:  pass,
+		Namespace: namespace,
+		Protocol:  strings.ToLower(string(infinispan.GetEndpointScheme())),
+	}
+	httpClient := curl.New(httpConfig, z.Kube)
+
+	if phase == ZeroPending {
+		return z.execute(httpClient, request, instance)
+	}
+	// Phase must be ZeroRunning, so wait for execution to complete
+	return z.waitForExecutionToComplete(httpClient, request, instance)
+}
+
+func (z *Controller) initializeResources(request reconcile.Request, instance Resource) (reconcile.Result, error) {
+	ctx := context.Background()
+	name := request.Name
+	namespace := request.Namespace
+	spec := instance.Spec()
+	statefulset := &appsv1.StatefulSet{}
+	clusterKey := k8client.ObjectKey{
+		Namespace: namespace,
+		Name:      spec.Cluster,
+	}
+	if err := z.Client.Get(ctx, clusterKey, statefulset); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{RequeueAfter: consts.DefaultWaitOnCluster}, fmt.Errorf("Unable to load Infinispan Cluster '%s': %w", spec.Cluster, err)
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// If a VolumeSource has not been configured, then we create a new pvc that will not be deleted with the CR
```

Comment:
What's the user story behind this? Is the leftover PVC and the inner archive going to be usable outside of the current OpenShift? If that's not the case then I think Operator should clean up after itself. If user deletes Backup with PVC created by Operator then I would expect that there is no backup to recover the cluster from.

Reply:
Good point, I think your suggested behaviour makes more sense.

Reply:
Ofcourse Operator shouldn't delete user provided VolumeSource but do we want to support custom VolumeSources? That could be a secret and I can't imagine backing up a cluster into a secret. Would be way simpler if we supported backing up only in the same type that we support for running Infinispan cluster (PVC generated by Operator, possibly `storageSize`, `accessMode` and `storageClassName`).

Reply:
Originally I was thinking that it would allow greater flexibility, but I think you're right that it's overkill. I agree the user should be able to optionally provide `storageSize` and `storageClassName`.

Does the `accessMode` need to be configurable? I think that `ReadWriteOnce` should always be sufficient.

Reply:
Unless it's OpenStack Manilla [1]. But since we are using RWO with Infinispan cluster we wouldn't be able to run the cluster with this PV either so I guess we can drop that one.

https://docs.openshift.com/container-platform/4.5/storage/understanding-persistent-storage.html#pv-access-modes_understanding-persistent-storage